homeassistant/components/weatherflow_cloud/__init__.py

```diff
@@ -2,33 +2,126 @@
-from .coordinator import WeatherFlowCloudDataUpdateCoordinator
+from .const import DOMAIN, LOGGER
+from .coordinator import (
+    WeatherFlowCloudDataCallbackCoordinator,
+    WeatherFlowCloudUpdateCoordinatorREST,
+)
 
 PLATFORMS: list[Platform] = [Platform.SENSOR, Platform.WEATHER]
 
 
+@dataclass
+class WeatherFlowCoordinators:
+    """Data Class for Entry Data."""
+
+    rest: WeatherFlowCloudUpdateCoordinatorREST
+    wind: WeatherFlowCloudDataCallbackCoordinator[
+        EventDataRapidWind, RapidWindListenStartMessage, RapidWindWS
+    ]
+    observation: WeatherFlowCloudDataCallbackCoordinator[
+        WebsocketObservation, ListenStartMessage, WebsocketObservation
+    ]
+
+
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up WeatherFlowCloud from a config entry."""
 
-    data_coordinator = WeatherFlowCloudDataUpdateCoordinator(
+    LOGGER.debug("Initializing WeatherFlowCloudDataUpdateCoordinatorREST coordinator")
+
+    rest_api = WeatherFlowRestAPI(
+        api_token=entry.data[CONF_API_TOKEN], session=async_get_clientsession(hass)
+    )
+
+    stations: StationsResponseREST = await rest_api.async_get_stations()
+
+    # Define Rest Coordinator
+    rest_data_coordinator = WeatherFlowCloudUpdateCoordinatorREST(
+        hass=hass, rest_api=rest_api, stations=stations
+    )
+
+    # Initialize the stations
+    await rest_data_coordinator.async_config_entry_first_refresh()
+
+    # Construct Websocket Coordinators
+    LOGGER.debug(
+        "Initializing WeatherFlowCloudDataUpdateCoordinatorWebsocketWind coordinator"
+    )
+    websocket_device_ids = rest_data_coordinator.device_ids
+
+    # Build API once
+    websocket_api = WeatherFlowWebsocketAPI(
+        access_token=entry.data[CONF_API_TOKEN], device_ids=websocket_device_ids
+    )
+
+    websocket_observation_coordinator = WeatherFlowCloudDataCallbackCoordinator[
+        WebsocketObservation, WebsocketObservation, ListenStartMessage
+    ](
+        hass=hass,
+        rest_api=rest_api,
+        websocket_api=websocket_api,
+        stations=stations,
+        listen_request_type=ListenStartMessage,
+        event_type=EventType.OBSERVATION,
+    )
+
+    websocket_wind_coordinator = WeatherFlowCloudDataCallbackCoordinator[
+        EventDataRapidWind, EventDataRapidWind, RapidWindListenStartMessage
+    ](
         hass=hass,
-        api_token=entry.data[CONF_API_TOKEN],
+        stations=stations,
+        rest_api=rest_api,
+        websocket_api=websocket_api,
+        listen_request_type=RapidWindListenStartMessage,
+        event_type=EventType.RAPID_WIND,
     )
-    await data_coordinator.async_config_entry_first_refresh()
 
-    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = data_coordinator
+    # Run setup method.
+    await asyncio.gather(
+        websocket_wind_coordinator._async_setup(),  # noqa: SLF001,
+        websocket_observation_coordinator._async_setup(),  # noqa: SLF001
+    )
```

Comment:
Let's not access the private parts

Reply:
> You can override _async_setup in your coordinator, and it will be automatically called during coordinator.async_config_entry_first_refresh(). It offers the same error handling as _async_update_data and will handle ConfigEntryError and ConfigEntryAuthFailed accordingly.

Went with this change as-per docs.

Reply:
Refactored to a public `async_setup` method